### PR TITLE
automatic mapping of table names to extractors

### PR DESCRIPTION
**Pull request by en-GB**
removes some manual steps to adding new shops

### CODE REVIEW

File: extract/extract/extractors/otto.py [hunk 1]
```diff
@@ -268,3 +269,6 @@ def _get_sustainability(product_data: dict, parsed_url: ParseResult) -> List[str
         labels.update(_get_sustainability_info(sustainable_soup))
 
     return sorted({_LABEL_MAPPING.get(label, LabelIDType.UNKNOWN) for label in labels.keys()})
+
+
+EXTRACTOR_FOR_TABLE_NAME = {TABLE_NAME_SCRAPING_OTTO: extract_otto}
```
Comment (se-jaeger) [inline extract/extract/extractors/otto.py:274]: We still need these mappings?

File: extract/extract/extractors/zalando.py [hunk 2]
```diff
@@ -133,3 +134,6 @@ def _get_sustainability(beautiful_soup: BeautifulSoup) -> List[str]:
         labels = cluster.find_all(attrs={"data-testid": "certificate__title"})
         return sorted({_LABEL_MAPPING.get(label.string, LabelIDType.UNKNOWN) for label in labels})
     return []
+
+
+EXTRACTOR_FOR_TABLE_NAME = {TABLE_NAME_SCRAPING_ZALANDO: extract_zalando}
```
Comment (se-jaeger) [inline extract/extract/extractors/zalando.py:139]: See above..

**Review by se-jaeger — COMMENTED** [review]
In the future, short description of the PR would be nice.

To be honest, I don't see the benefit here. If this mapping is split and moved to separate files and then created dynamically, IMHO it makes things worse than have a single and static mapping.

Is it possible to "find" these functions (I think it's also fine to expect that this function exists), e.g., `extract_otto` in the file `otto` dynamically and directly return the mapping instead of return the `names`?

File: extract/extract/extractors/__init__.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# flake8: noqa
+
+import pkgutil
+
+__all__ = [module.name for module in pkgutil.iter_modules(__path__)]
+from . import *
```
[the remark this answers is not in the document]
Comment (en-GB) [reply extract/extract/extractors/__init__.py:6]: it seems to be.
i dont know that much about python modules so im sure there is a better way to do this

**Comment by se-jaeger** [comment]
I pushed a suggestions. It does not need any indirections or decorators and has much fewer lines of code.
It just relies on file names, e.g., `otto.py` or `zalando.py` used as table names and they need to expose the function `extract` (no suffix). WDYT?

**Comment by en-GB** [comment]
by indirection you mean the `import *` in `extractors/__init__.py`?
i still feel like thats better than using `import_module`

File: extract/extract/extractors/__init__.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import pkgutil
+
+names = __all__ = [module_info.name for module_info in pkgutil.iter_modules(__path__)]  # noqa: 405
+# this doesnt import anything.
+
+from . import *  # noqa: 403
+
+# this actually imports the extractors
+# for some reason it only finds them when __all__ is set
```
Comment (se-jaeger) [inline extract/extract/extractors/__init__.py:9]: Can we construct `EXTRACTOR_FOR_TABLE_NAME` in this file?

File: extract/extract/__init__.py [hunk 5]
```diff
@@ -1,21 +1,16 @@
...
 
 from core import log
-from core.constants import TABLE_NAME_SCRAPING_OTTO, TABLE_NAME_SCRAPING_ZALANDO
 from core.domain import Product, ScrapedPage
 
-# Because we ignored the files `zalando.py` and `otto.py` we have to skip them here as well
-from .extractors.otto import extract_otto  # type: ignore[attr-defined]
-from .extractors.zalando import extract_zalando  # type: ignore[attr-defined]
-from .parse import parse_page
+from . import extractors
```
Comment (se-jaeger) [inline extract/extract/__init__.py:6]: And then import it here: `from .extractors import EXTRACTOR_FOR_TABLE_NAME`?

**Review by se-jaeger — CHANGES_REQUESTED** [review]
I would prefer to be as explicit as possible. Using `import *` imports things we do not know about. If possible, let avoid it.

**Comment by en-GB** [comment]
closed as discussed

**Pull request closed without merging**